design-documents/asynchronous-import/modularity/source-data-retrieving.md

```diff
@@ -0,0 +1,153 @@
+
+## Implementation
+
+### API
+
+```php
+namespace Magento\AsynchronousImportSourceDataRetrievingApi\Api;
+
+use Magento\AsynchronousImportSourceDataRetrievingApi\Api\Data\SourceInterface;
+use Magento\AsynchronousImportSourceDataRetrievingApi\Api\Data\SourceDataInterface;
+use Magento\Framework\Validation\ValidationException;
+
+/**
+ * Retrieve source data operation. Uses differect strategies for source data retrieving
+ *
+ * @api
+ */
+interface RetrieveSourceDataInterface
+{
+    /**
+     * Retrieve source data operation. Uses differect strategies for source data retrieving
+     *
+     * @param SourceInterface $source
+     * @return SourceDataInterface
+     * @throws ValidationException
+     * @throws SourceDataRetrievingException
+     */
+    public function execute(SourceInterface $source): SourceDataInterface;
+}
+```
+
+```php
+namespace Magento\AsynchronousImportSourceDataRetrievingApi\Api\Data;
+
+/**
+ * Describes how to retrieve data from data source
+ *
+ * @api
+ */
+interface SourceInterface
+{
+    /**
+     * Get source type
+     *
+     * @return string
+     */
+    public function getSourceType(): string;
+
+    /**
+     * Get source definition
+     *
+     * @return string
+     */
+    public function getSourceDefinition(): string;
+
+    /**
+     * Get source data format
+     *
+     * @return string
+     */
+    public function getSourceDataFormat(): string;
+}
+```
+
+```php
+namespace Magento\AsynchronousImportSourceDataRetrievingApi\Api\Data;
+
+/**
+ * Represents retrieved source data (result of retrieving operation)
+ *
+ * @api
+ */
+interface SourceDataInterface extends \IteratorAggregate
+{
+    public const ITERATOR = 'iterator';
+}
+```
+
+### Extension points
+
+```php
+namespace Magento\AsynchronousImportSourceDataRetrievingApi\Model;
+
+use Magento\AsynchronousImportSourceDataRetrievingApi\Api\Data\SourceInterface;
+use Magento\AsynchronousImportSourceDataRetrievingApi\Api\SourceDataRetrievingException;
+
+/**
+ * Extension point for adding source data retrieving algorithms
+ * Represents concrete strategy
+ *
+ * @api
+ */
+interface RetrieveSourceDataStrategyInterface
+{
+    /**
+     * Source data retrieving strategy
+     *
+     * @param SourceInterface $source
+     * @return \Traversable
+     * @throws SourceDataRetrievingException
+     */
+    public function execute(SourceInterface $source): \Traversable;
+}
+```
+
+```php
+namespace Magento\AsynchronousImportSourceDataRetrievingApi\Model;
+
+/**
+ * Extension point for adding source validators via DI configuration
+ *
+ * @api
+ */
+class SourceValidatorChain implements SourceValidatorInterface
+{
+  ...
+}
+```
+
+```php
+namespace Magento\AsynchronousImportSourceDataRetrievingApi\Model;
+
+use Magento\AsynchronousImportSourceDataRetrievingApi\Api\Data\SourceInterface;
+use Magento\Framework\Validation\ValidationResult;
+
+/**
+ * Extension point for adding source validators
+ *
+ * @api
+ */
+interface SourceValidatorInterface
```

Comment:
How & where we will going to use this extension point?

Reply:
It will be used: 
https://github.com/magento/architecture/pull/330/files#diff-610ea2fd7eb8081598349a79459e9313R40
On a step we are retrieving data to validate them

Reply:
is it injected through DI configuration?

Reply:
right